src/lifecycle/index.ts

```diff
@@ -0,0 +1,512 @@
+import { Dispatch, useEffect, useReducer } from 'react';
+import { QueryableConsts, QueryableStorage } from '@polkadot/api/types';
+import { getAllReferenda, getAllTracks } from '../chain/referenda';
+import { DEFAULT_NETWORK, endpointsFor, Network, networkFor } from '../network';
+import { AccountVote, Referendum, ReferendumOngoing } from '../types';
+import { err, ok, Result } from '../utils';
+import { all, open, save } from '../utils/indexeddb';
+import { measured } from '../utils/performance';
+import { newApi } from '../utils/polkadot-api';
+import { timeout } from '../utils/promise';
+import {
+  Action,
+  ChainState,
+  PersistedDataContext,
+  Report,
+  State,
+} from './types';
+
+// Auto follow chain updates? Only if user settings? Show notif? Only if impacting change?
+// Revisit if/when ChainState is persisted
+
+// TODO also load from polkassembly and others
+// https://github.com/evan-liu/fetch-queue
+// https://github.com/jkramp/fetch-queue
+// https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide
+
+// Actions
+// TODO
+// Heartbeat (to detect no new finalized blocks are reeived, keep track of delta between last received and now)
+
+// Startup
+// settings, overriden by Env variable, overriden by query params, overriden by user settings? offer user confirmation
+// network, allow to restore chain db
+// endpoints, allow to connect to chain
+// if no endpoints provided, default to network endpoints
+// endpoints and network: must match network
+// only endpoints: sets network
+
+/**
+ * @param referenda
+ * @returns a subset containing only ReferendumOngoing
+ */
+export function filterOngoingReferenda(
+  referenda: Map<number, Referendum>
+): Map<number, ReferendumOngoing> {
+  return new Map(
+    [...referenda].filter(([, v]) => v.type == 'ongoing') as [
+      number,
+      ReferendumOngoing
+    ][]
+  );
+}
+
+/**
+ * @param referenda
+ * @param votes
+ * @returns a subset of referenda not yet voted on
+ */
+export function filterToBeVotedReferenda(
+  referenda: Map<number, ReferendumOngoing>,
+  votes: Map<number, AccountVote>
+): Map<number, ReferendumOngoing> {
+  return new Map([...referenda].filter(([index]) => !votes.has(index)));
+}
+
+function withNewReport(previousState: State, report: Report): State {
+  const previousReports = previousState.reports || [];
+  return {
+    ...previousState,
+    reports: [report, ...previousReports],
+  };
+}
+
+function withFailedTransition(previousState: State): State {
+  return withNewReport(previousState, {
+    type: 'Error',
+    message: `Incorrect transition: ${previousState.type}`,
+  });
+}
+
+function reducer(previousState: State, action: Action): State {
+  // Note that unlucky timing might lead to overstepping changes (triggered via listeners registered just above)
+  // So applying changes must be indempotent
+  switch (action.type) {
+    case 'SetConnectedAccountAction': {
```

Comment:
At somepoint would be good to set an enum for actions to avoid inconsistencies, mistakes.

Reply:
There can't be inconsistencies, this is enforced by the typescript compiler. See https://basarat.gitbook.io/typescript/type-system/discriminated-unions

Reply:
Cool! I missed the switch is on types